stack/stack.mbt

```diff
@@ -0,0 +1,428 @@
+///
+/// # Example
+///
+/// ```
+/// let s: Stack[Int] = Stack::from_array([1,2,3])
+/// let s1: Stack[Int] = Stack::new()
+/// println(s.pop()) // output: Some(1)
+/// println(s) // output: { elements: Cons(2, Cons(1, Nil)), len: 2 }
+/// println(s1.pop()) // output: None
+/// ```
+pub fn pop[T](self : Stack[T]) -> Option[T] {
+  match self.elements {
+    Cons(hd, tl) => {
+      self.elements = tl
+      self.len = self.len - 1
+      Some(hd)
+    }
+    Nil => None
+  }
+}
+
+test "pop" {
+  let s : Stack[Int] = Stack::[1, 2, 3]
+  let s1 : Stack[Int] = Stack::new()
+  @assertion.assert_eq(s.pop(), Some(3))?
+  @assertion.assert_eq(s.elements, Cons(2, Cons(1, Nil)))?
+  @assertion.assert_eq(s.len, 2)?
+  @assertion.assert_eq(s1.pop(), None)?
+  @assertion.assert_eq(s1.elements, Nil)?
+  @assertion.assert_eq(s1.len, 0)?
+}
+
+/// Drop the element at the top of the stack.
+/// Like pop, but does not return elements and does nothing if the Stack is empty.
+///
+/// # Example
+///
+/// ```
+/// let s: Stack[Int] = Stack::from_array([1,2,3])
+/// s.drop()
+/// println(s) // output: { elements: Cons(2, Cons(3, Nil)), len: 2 }
+/// ```
+pub fn drop[T](self : Stack[T]) -> Unit {
+  match self.elements {
+    Cons(_hd, tl) => {
+      self.elements = tl
+      self.len = self.len - 1
+    }
+    Nil => ()
+  }
+}
+
+test "drop" {
+  let s : Stack[Int] = Stack::[1, 2, 3]
+  s.drop()
+  @assertion.assert_eq(s.elements, Cons(2, Cons(1, Nil)))?
+  @assertion.assert_eq(s.len, 2)?
+}
+
+/// Only the top element of the stack is returned and will not be pop or drop.
+/// If there are elements in the stack, return `Some (the top element of the stack)`, otherwise return `None`.
+///
+/// # Example
+///
+/// ```
+/// let s: Stack[Int] = Stack::from_array([1,2,3])
+/// println(s.top()) // output: Some(3)
+/// println(s) // output: { elements: Cons(3, Cons(2, Cons(1, Nil))), len: 3 }
+/// ```
+pub fn top[T](self : Stack[T]) -> Option[T] {
+  match self.elements {
+    Cons(hd, _) => Some(hd)
+    Nil => None
+  }
+}
+
+test "top" {
+  let s : Stack[Int] = Stack::[1, 2, 3]
+  @assertion.assert_eq(s.top(), Some(3))?
+  @assertion.assert_eq(s.elements, Cons(3, Cons(2, Cons(1, Nil))))?
+  @assertion.assert_eq(s.len, 3)?
+}
+
+/// If stack is empty, return true, otherwise return false.
+///
+/// # Example
+///
+/// ```
+/// println(Stack::from_array([1, 2, 3]).is_empty()) // output: false
+/// println(Stack::new().is_empty()) // output: true
+/// ```
+pub fn is_empty[T](self : Stack[T]) -> Bool {
+  self.len == 0
+}
+
+test "is_empty" {
+  let empty : Stack[Unit] = Stack::new()
+  @assertion.assert_false(Stack::[1, 2, 3].is_empty())?
+  @assertion.assert_true(empty.is_empty())?
+}
+
+/// Returns the number of elements of the Stack
+pub fn length[T](self : Stack[T]) -> Int {
+  self.len
+}
+
+// Compares whether the elements in two lists are the same
+// pub fn equal[T](self: Stack[T], other: Stack[T]) -> Bool {
```

Comment:
Did you forget to uncomment this code? The method name of trait `Eq` is `op_equal`